devicetree documentation improvements for v3.3

mbolivar-nordic opened the pull request:
Generic improvements and rework only. This does **not** include release notes, which will arrive during the stabilization period.

mbolivar-nordic committed:
doc: dts: split up bindings docs

The bindings.rst page has gotten too big. Split it up into
sub-documents to improve readability and restore maintainability.

As part of this effort, move the /zephyr,user node documentation into
its own page in the guide. This page has proven extremely difficult
for users to spot in my experience, and it's meant as a convenience,
so let's make the documentation more convenient as well.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: bindings-syntax improvements

Generic improvements:

- clean up some language that needs adjusting
- rename some section headers and reorder some content to
  improve readability
- add a table of contents to ease search

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: split up intro

The devicetree introduction page is too big. Split it up to improve
readability and restore maintainability. Add more section headers and
do some other rearranging now that it's more convenient to do so.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: add guide to phandles

Phandles, specifier spaces, and cell names are simultaneously
extremely common and woefully underdocumented. Address that by:

- reworking our existing documentation on these subjects in
  bindings-syntax.rst, fixing missing information in the
  property syntax template as well

- adding a standalone guide which describes how all the pieces
  fit together, providing a bridge for the gap between
  DTS/bindings and C APIs

My goal is not to eventually make this a comprehensive place
where *all* specifier spaces are documented. It would be better (more
scalable, more discoverable) to improve the individual API pages to
cover the devicetree-related conventions that apply in each case.
That's a problem for someone else and another day, but we do need a
few concrete references in the DTS guides to keep the motivation
clear.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: clarify a design requirement

There's been some confusion about what we mean by hardware
description. We're really talking about the user-facing inputs that
need to be configured per SoC, application, etc. It's fine to do
things like use ifdeffery on a CMSIS header in an aarch32 support file
to decide if the current target has some feature, for instance -- that
sort of thing doesn't *have* to come from DT.

At the same time, we don't want to encourage vendor-specific hardware
configuration languages from creeping into upstream zephyr, so keep
the language strong in an effort to avoid that misinterpretation.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: improve zephyr-user-node.rst

Now that this information is in a separate page instead of buried at
the bottom of the DT bindings documentation, it's more convenient to
split it up into subsections for readability.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: improve bindings-upstream

Splitting up the 'general rules' section into subsections makes it
possible to link directly to a particular rule. This is useful when
pointing out an issue during code review, sharing with another
colleague, etc. Adding a local table of contents makes the page
skimmable (it's buried too deeply in the toctree to have sections
listed in the HTML sidebar).

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: document zephyr, prefix for props/compats

We have established a convention that the zephyr, prefix in property
names indicates some sort of zephyr-specific extension to a common
binding, or a zephyr-specific driver configuration knob. We also have
established a convention that compatibles which begin with "zephyr,"
are specific to our operating system. Document these facts.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: touch up api-usage

Just a few minor things from a periodic review.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: touch up api

Just a few minor things from a periodic review.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

nordicjm commented on doc/build/dts/bindings-syntax.rst:
```diff
@@ -0,0 +1,799 @@
+For rules related to ``default`` in upstream Zephyr bindings, see
+:ref:`dt-bindings-default-rules`.
+
+See :ref:`dt-bindings-example-properties` for examples. Putting ``default:`` on
+any property type besides those used in the examples will raise an error.
```
~~Is this enforced? I ask as for a hd44780 driver I am using this for the main binding itself for line addresses, most displays use the same addresses but on the odd chance that there is one somewhere that doesn't, it allows it to be overridden without the majority of displays needing to put common values in here.~~

~~Just had a look through in-tree bindings and it's used in a lot of places (quite a few in displays)~~
After re-reading it a number of times, I see that it means refer to the linked examples to see which types you are allowed to use default on, not that you can only use it in example bindings for sample projects. This text could be clarified a bit better, or list the types allowed here in the text.

nordicjm commented on doc/build/dts/intro-syntax-structure.rst:
```diff
@@ -0,0 +1,472 @@
+    Information used to address the device. The value is specific to the device
+    (i.e. is different depending on the compatible property).
+
+    The ``reg`` property is a sequence of ``(address, length)`` pairs. Each
```
Worth mentioning that it is in hex, e.g. for i2c if someone is looking at a datasheet, they might not be aware that they need to put in 1B and not 27

nordicjm commented on doc/build/dts/intro-syntax-structure.rst:
```diff
@@ -0,0 +1,472 @@
+   :header-rows: 1
+   :widths: 1 4 4
+
+   * - Property type
```
Duplicate of what is on the bindings-syntax page (which has additional rows)

mbolivar-nordic replied:
Yep, that's on purpose. `path` and `compound` aren't worth covering in the introduction in my opinion.

mbolivar-nordic commented on doc/build/dts/bindings-intro.rst:
```diff
@@ -0,0 +1,134 @@
+.. _dt-binding-compat:
```
I don't understand what you mean by "top level page".

The reason why these pages don't show up in the left hand sidebar is because the DT pages are buried too deeply. That's why I added `..contents::` directives to several of them where I anticipate internal navigation being appreciated by readers.

mbolivar-nordic commented on doc/build/dts/bindings-intro.rst:
```diff
@@ -0,0 +1,134 @@
+
+- the zephyr repository
+- your :ref:`application source directory <application>`
+- your :ref:`board directory <board_porting_guide>`
```
Can you cite your source?

nordicjm replied:
cmake/modules/dts.cmake line 51:

```
# 'DTS_ROOT' is a list of directories where a directory tree with DT
# files may be found. It always includes the application directory,
# the board directory, and ${ZEPHYR_BASE}.
list(APPEND
  DTS_ROOT
  ${APPLICATION_SOURCE_DIR}
  ${BOARD_DIR}
  ${SHIELD_DIRS}
  ${ZEPHYR_BASE}
  )
```

mbolivar-nordic replied:
Interesting that this is done but unused upstream (or in NCS). Added now as a separate commit since this gap in the docs existed before this PR (you're seeing it in green because this content was moved from another file). Thanks!

mbolivar-nordic committed:
doc: dts: clarify reg format

These are in hex.

Suggested-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@nordicjm thanks for review, comments addressed

henrikbrixandersen commented on doc/build/dts/bindings-upstream.rst:
```diff
@@ -38,49 +41,59 @@ In particular, this rule applies if:
-  uppercase. Use dashes (``-``) instead of underscores. (The one exception to
-  this rule is if you are replicating a well-established binding from somewhere
-  like Linux.)
+Do not use uppercase letters (``A`` through ``Z``) or underscores (``_``) in
+property names. Use lowercase letters (``a`` through ``z``) instead of
+uppercase. Use dashes (``-``) instead of underscores. (The one exception to
+this rule is if you are replicating a well-established binding from somewhere
+like Linux.)
```
Should we recommend using all-lowercase for enum property values as well?

mbolivar-nordic replied:
Is there a linux convention for that?

henrikbrixandersen replied:
I believe so, but I am have not been able to find any documentation for it.

mbolivar-nordic replied:
I think I'd rather leave it out for now, then, since it's ambiguous.

henrikbrixandersen commented on doc/build/dts/bindings-upstream.rst:
```diff
@@ -208,3 +208,28 @@ Some examples of what **not** to do, and why:
+  is the ``zephyr,lazy-load`` property in the :dtcompatible:`ti,bq274xx`
+  binding. Though devicetree in general is a hardware description and
+  configuration language, it is Zephyr's only mechanism for configuring driver
+  behavior for individual devices. Therefore, as a compromise, we do allow some
```
Perhaps replace "individual devices" with "individual device driver instances"?

mbolivar-nordic replied:
Personally, I have never been a fan of calling each `struct device` a "device driver instance". IMO this is a longstanding wart in our device driver model documentation. It's a `struct device`, so I try to be consistent about calling them "devices".

henrikbrixandersen replied:
My problem with using just "devices" is that it is ambiguous - at least to me. Not a show-stopper.

mbolivar-nordic replied:
OK, I think this should suit us both:

```
behavior for an individual ``struct device``
```

henrikbrixandersen reviewed (dismissed):
Looks good. Very comprehensive as always. A couple of minors.

jfischer-no reviewed (commented):
LGTM

mbolivar-nordic committed:
doc: dts: add missing entry to dts/bindings search path

Since SHIELD_DIRS gets added to DTS_ROOT in dts.cmake, any shield
directories are also places where we look for bindings by default.
This feature is not used in upstream Zephyr, but it is supported,
so document it.

Suggested-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Please re-approve, I added one commit to address review from @nordicjm
